stempeg/read.py

```diff
@@ -1,200 +1,201 @@
+            metadata = info
+
+        ffmpeg.probe(filename)
+
+    except ffmpeg._run.Error as e:
+        raise Warning(
+            'An error occurs with ffprobe (see ffprobe output below)\n\n{}'
+            .format(e.stderr.decode()))
+    if 'streams' not in metadata.info or metadata.nb_audio_streams == 0:
+        raise Warning('No stream was found with ffprobe')
+
+    if stems_from_channels:
+        if metadata.nb_audio_streams != 1:
+            raise Warning(
+                'In this configuration, only a single substream is processed'
+            )
+        else:
+            if metadata.audio_streams[0][
+                'channels'
+            ] % stems_from_channels != 0:
+                raise Warning('Stems should be encoded as multi-channel')
+            else:
+                substreams = 0
+    else:
+        if stem_id is not None:
+            substreams = stem_id
+        else:
+            substreams = metadata.audio_stream_idx()
 
-    # 12 digits should be enough to represent a single sample of
-    # 192khz in float
-    ctx.prec = precision
+    if not isinstance(substreams, list):
+        substreams = [substreams]
 
-    d1 = ctx.create_decimal(repr(round(f, 5)))
-    return format(d1, 'f')
+    stems = []
+    # apply fix for very small start values of `-ss <0.000001`
+    if start:
+        if start < 1e-4:
+            start = None
+
+    for stem in substreams:
+        if sample_rate is None:
+            sample_rate = metadata.sample_rate(stem)
+        channels = metadata.channels(stem)
+        output_kwargs = {'format': 'f32le', 'ar': sample_rate}
+        if duration is not None:
+            output_kwargs['t'] = _to_ffmpeg_time(duration)
+        if start is not None:
+            output_kwargs['ss'] = _to_ffmpeg_time(start)
+
+        output_kwargs['map'] = '0:' + str(stem)
+        process = (
+            ffmpeg
+            .input(filename)
+            .output('pipe:', **output_kwargs)
+            .run_async(pipe_stdout=True, pipe_stderr=True))
+        buffer, _ = process.communicate()
+        waveform = np.frombuffer(buffer, dtype='<f4').reshape(-1, channels)
+        if not waveform.dtype == np.dtype(dtype):
+            waveform = waveform.astype(dtype)
+        stems.append(waveform)
+
+    stem_durations = np.array([t.shape[0] for t in stems])
+    if not (stem_durations == stem_durations[0]).all():
+        warnings.warning("Stems differ in length and were shortend")
+        min_length = np.min(stem_durations)
+        stems = [t[:min_length, :] for t in stems]
+        # TODO: Also check number of channels are the same for all
+
+    stems = np.array(stems)
+    if stems_from_channels and stems.shape[-1] > 1:
+        stems = stems.transpose(1, 0, 2)
+        stems = stems.reshape(
+            stems.shape[0], stems.shape[1], -1, stems_from_channels
+        )
+        stems = stems.transpose(2, 0, 3, 1)[..., 0]
+
+    if not always_3d:
+        stems = np.squeeze(stems)
```

Comment:
This will also remove the last dimension in the mono case (data would have (N,) shape) is it the intended behavior ?
```
S, rate = stempeg.read_stems("test_file_mono.wav")
S.shape```

Reply:
yes, I think that's what I would expect. I could an additional `always_2d` flag like in `soundfile` that would make sure that its at_least 2d. Would you like that?